### PR TITLE
PA and STRAC onboarding using Redox

### PR DESCRIPTION
**Pull request by RickHawesUSDS**
This PR adds the STRAC app sending to PA using the Redox engine. To make this path work, a number of new features were added including retry logic. 

## Changes 
- Retry logic for Redox and SFTP transports
- Redox serializers and transports
- Multi-level schema inheritance
- A mock-server for Redox

## Checklist
- [x] Tested locally?
- [ ] Added new dependencies?
- [x] Updated the docs?
- [x] Added a test?
- [ ] Did you check for sensitive data, and remove any?
- [ ] Are additional approvals needed for this change?
- [ ] Are there potential vulnerabilities or licensing issues with any new dependencies introduced?

## Fixes 

- #228 

## To Be Done

- Redox configuration still needs to be worked through.

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -85,3 +87,11 @@ jobs:
...
               --name ${{ env.RESOURCE_GROUP }} \
               --settings PRIME_ENVIRONMENT=${{ env.PRIME_ENVIRONMENT }} ${{ env.SFTP_CREDS }}
+
+      - name: Set Redox Environment Variables
+        run: |
+              az functionapp config appsettings set \
+              --resource-group ${{ env.RESOURCE_GROUP }} \
+              --name ${{ env.RESOURCE_GROUP }} \
+              --settings REDOX_SECRET=${{ env.REDOX_SECRET }} 
+
```
Comment (RickHawesUSDS) [inline .github/workflows/release.yml:97]: The REDOX secret is kept in KeyBase as well.

File: prime-router/docker-compose.yml [hunk 2]
```diff
@@ -50,4 +51,15 @@ services:
...
+    image: mockserver/mockserver:mockserver-5.11.1
+    ports:
+      - 1080:1080
+    environment:
+      MOCKSERVER_PROPERTY_FILE: /config/mockserver.properties
+      MOCKSERVER_INITIALIZATION_JSON_PATH: /config/initializerJson.json
+    volumes:
+      - type: bind
+        source: ./src/test/redox
+        target: /config
```
Comment (RickHawesUSDS) [inline prime-router/docker-compose.yml:65]: This mock server emulates the Redox end-point.

File: prime-router/metadata/organizations-local.yml [hunk 3]
```diff
@@ -91,38 +114,21 @@
...
-  # HHS
-  - name: fed1
-    description: Federal feed
-    services:
-      - name: elr
-        topic: sample
-        schema: sample/fed1-sample
-        jurisdictionalFilter: [ "matches(observation, covid-19:.*)", "matches(state, .*)"]
-        transforms: {deidentify: true}
-        format: CSV
```
Comment (RickHawesUSDS) [inline prime-router/metadata/organizations-local.yml:128]: Removing these test configs. Not needed.

File: prime-router/metadata/organizations-prod.yml [hunk 4]
```diff
@@ -10,6 +10,14 @@
...
 
+  - name: strac
+    description: STRAC POC testing app
+    clients:
+      - name: default
+        topic: covid-19
+        schema: strac/strac-covid-19
+        format: CSV
+
   - name: az-phd
```
Comment (RickHawesUSDS) [inline prime-router/metadata/organizations-prod.yml:21]: STRAC has a new client and new schema.

File: prime-router/metadata/organizations-prod.yml [hunk 5]
```diff
@@ -65,38 +73,21 @@
...
+#        transforms: { deidentify: false }
+#        defaults:
+#          redox_destination_id: cd019512-8a58-40e5-a416-4e597404f9b8
+#          redox_destination_name: "CDC Pennsylvania Dept of Health Destination (s)"
+#          redox_source_id: d89d4057-930f-4c5b-bb39-8cddb326e928
+#          redox_source_name: "Prime Data Hub (Staging)"
+#        format: REDOX
+#        transports:
+#          - type: REDOX
+#            apiKey: 9e7f9bd1-a876-4f10-9d17-0833dc7bf60d # Redox prod
```
Comment (RickHawesUSDS) [inline prime-router/metadata/organizations-prod.yml:93]: The PROD version is not ready.

File: prime-router/metadata/organizations-test.yml [hunk 6]
```diff
@@ -65,38 +73,35 @@
...
+        topic: covid-19
+        schema: co/co-covid-19-redox
+        jurisdictionalFilter: [ "matches(ordering_facility_state, CO)" ]
+        transforms: { deidentify: false }
+        defaults:
+          redox_destination_id: 62d62d52-3771-4151-aff4-4a3d420a8b7a
+          redox_destination_name: "Colorado Dept of Public Health (CDPHE)"
+          redox_source_id: d89d4057-930f-4c5b-bb39-8cddb326e928
+          redox_source_name: "Prime Data Hub (Staging)"
+        format: REDOX
```
Comment (RickHawesUSDS) [inline prime-router/metadata/organizations-test.yml:89]: CO is setup in staging.

File: prime-router/metadata/schemas/covid-19.schema [hunk 7]
```diff
@@ -161,7 +161,7 @@ elements:
     natFlatFileField: Hospitalized
     hhsGuidanceField:
     hl7Field: AOE
-    hl7AOEQuestion: 7974-4
+    hl7AOEQuestion: 77974-4
     documentation: Is the patient hospitalized?
```
Comment (RickHawesUSDS) [inline prime-router/metadata/schemas/covid-19.schema:165]: Correct an error

File: prime-router/metadata/schemas/covid-19.schema [hunk 8]
```diff
@@ -178,7 +178,7 @@ elements:
     natFlatFileField: Illness_onset_date
     hhsGuidanceField:
     hl7Field: AOE
-    hl7AOEQuestion: 11368-8
+    hl7AOEQuestion: 65222-2
```
Comment (RickHawesUSDS) [inline prime-router/metadata/schemas/covid-19.schema:182]: Revised HHS guidance

File: prime-router/metadata/schemas/covid-19.schema [hunk 9]
```diff
@@ -227,7 +231,6 @@ elements:
     type: TABLE
     table: fips-county
     tableColumn: County
-    cardinality: ONE
     natFlatFileField: Ordering_facility_county
```
Comment (RickHawesUSDS) [inline prime-router/metadata/schemas/covid-19.schema:234]: Generally, removing cardinality ONE from elements of the main schema.

File: prime-router/metadata/schemas/covid-19.schema [hunk 10]
```diff
@@ -443,10 +452,14 @@ elements:
...
 
+  - name: patient_country
+    type: TEXT
+    default: USA
+
   - name: patient_email
     type: EMAIL
     cardinality: ZERO_OR_ONE
-    hl7Field: ORC-23-4
+    hl7Field: PID-13-4
```
Comment (RickHawesUSDS) [inline prime-router/metadata/schemas/covid-19.schema:463]: Another mistake in the HL7 schema.

File: prime-router/metadata/schemas/covid-19.schema [hunk 11]
```diff
@@ -474,6 +487,11 @@ elements:
 
       Other states may choose to define their own formats.
 
+  - name: patient_drivers_license
+    type: ID_DLN
+    cardinality: ZERO_OR_ONE
+    hl7Field: PID-20-1
+
```
Comment (RickHawesUSDS) [inline prime-router/metadata/schemas/covid-19.schema:494]: New element that is used by STRAC.

File: prime-router/metadata/valuesets/covid-19.valuesets [hunk 12]
```diff
@@ -109,16 +113,16 @@
   values:
     - code: 95417-2
       display: First test for condition of interest
-      version: 2.69-pre
+      version: 2.69
     - code: 95418-0
```
Comment (RickHawesUSDS) [inline prime-router/metadata/valuesets/covid-19.valuesets:117]: LONIC 2.69 is out.

File: prime-router/metadata/valuesets/covid-19.valuesets [hunk 13]
```diff
@@ -142,6 +146,9 @@
     - code: 30525-0
       display: Age
       version: 2.68
+    - code: 77974-4
+      display: Whether patient was hospitalized because of this condition
+      version: 2.68
```
Comment (RickHawesUSDS) [inline prime-router/metadata/valuesets/covid-19.valuesets:152]: Missing part of the valueset.

File: prime-router/pom.xml [hunk 14]
```diff
@@ -16,7 +16,7 @@
         <kotlin.code.style>official</kotlin.code.style>
         <kotlin.compiler.jvmTarget>11</kotlin.compiler.jvmTarget>
         <kotlin.compiler.incremental>false</kotlin.compiler.incremental>
-        <kotlin.version>1.4.10</kotlin.version>
+        <kotlin.version>1.4.20</kotlin.version>
         <serialization.version>1.0.1</serialization.version>
```
Comment (RickHawesUSDS) [inline prime-router/pom.xml:20]: This fixes a bug where `isEmpty` was marked as deprecated.

File: prime-router/pom.xml [hunk 15]
```diff
@@ -447,13 +447,32 @@
...
+            <groupId>org.json</groupId>
+            <artifactId>json</artifactId>
+            <version>20201115</version>
+        </dependency>
+        <dependency>
+            <groupId>com.hierynomus</groupId>
+            <artifactId>sshj</artifactId>
+            <version>0.27.0</version>
+        </dependency>
         <dependency>
```
Comment (RickHawesUSDS) [inline prime-router/pom.xml:476]: FUEL is a Kotlin based HTTP package.

File: prime-router/src/main/kotlin/FakeReport.kt [hunk 16]
```diff
@@ -58,10 +58,15 @@ class FakeReport(val metadata: Metadata) {
...
                 when (element.name) {
                     "specimen_source_site_code" -> "71836000"
+                    "test_result_status" -> randomChoice("F", "C")
                     else -> {
-                        val valueSet = element.valueSetRef
-                            ?: error("ValueSet ${element.valueSet} is not available}")
-                        val possibleValues = valueSet.values.map { it.code }.toTypedArray()
+                        val possibleValues = if (element.altValues?.isNotEmpty() == true) {
+                            element.altValues.map { it.code }.toTypedArray()
+                        } else {
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/FakeReport.kt:65]: Alt values for faker

File: prime-router/src/main/kotlin/JurisdictionalFilters.kt [hunk 17]
```diff
@@ -67,7 +67,7 @@ class DoesNotMatch : JurisdictionalFilter {
     override fun getSelection(args: List<String>, table: Table): Selection {
         if (args.size < 2) error("Expecting two or more args to filter $name:  (columnName, value, value, ...)")
         val columnName = args[0]
-        val pattern = args[1]
+        // val pattern = args[1]
         val values = args.subList(1, args.size)
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/JurisdictionalFilters.kt:71]: pattern wasn't being used and the compiler warned about it.

File: prime-router/src/main/kotlin/Mappers.kt [hunk 18]
```diff
@@ -92,7 +96,15 @@ class UseMapper : Mapper {
         return if (values.isEmpty()) {
             null
         } else {
-            values.first().value
+            val (fromElement, fromValue) = values.first()
+            when {
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Mappers.kt:100]: added a couple more cases to use.

File: prime-router/src/main/kotlin/Report.kt [hunk 19]
```diff
@@ -187,43 +195,60 @@ class Report {
...
-    private fun buildColumn(mapping: Translator.Mapping, toElement: Element): StringColumn {
+    private fun buildColumnPass1(mapping: Translator.Mapping, toElement: Element): StringColumn? {
         return when (toElement.name) {
             in mapping.useDirectly -> {
                 table.stringColumn(mapping.useDirectly[toElement.name]).copy().setName(toElement.name)
             }
             in mapping.useMapper -> {
-                createMappedColumn(toElement, mapping.useMapper.getValue(toElement.name))
+                null
             }
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Report.kt:211]: Moved to multi-pass algorithm to handle mappers with default values on input.

File: prime-router/src/main/kotlin/Schema.kt [hunk 20]
```diff
@@ -43,7 +43,30 @@ data class Schema(
...
+        trackingElement = trackingElement,
+        description = description,
+        referenceUrl = referenceUrl,
+        extends = extends,
+        extendsRef = null,
+        basedOn = basedOn,
+        basedOnRef = null,
+    )
+
     val baseName: String get() = formBaseName(name)
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/Schema.kt:70]: Useful for test cases.

File: prime-router/src/main/kotlin/azure/Event.kt [hunk 21]
```diff
@@ -84,7 +111,7 @@ class ReportEvent(action: Action, val reportId: UUID, at: OffsetDateTime? = null
 }
 
 class ReceiverEvent(action: Action, val receiverName: String, at: OffsetDateTime? = null) : Event(action, at) {
-    override fun toMessage(): String {
+    override fun toQueueMessage(): String {
         val afterClause = if (at == null) "" else "$messageDelimiter${DateTimeFormatter.ISO_DATE_TIME.format(at)}"
```
Comment (RickHawesUSDS) [inline prime-router/src/main/kotlin/azure/Event.kt:115]: Better name.

File: prime-router/src/test/redox/initializerJson.json [hunk 22]
```diff
@@ -0,0 +1,24 @@
+[
```
Comment (RickHawesUSDS) [inline prime-router/src/test/redox/initializerJson.json:1]: This is the mock-server configured to look like Redox.

**Review by jimduff-usds — CHANGES_REQUESTED** [review]
Just noting these errors:

1. Unit test failed:  `[ERROR] gov.cdc.prime.router.TranslatorTests.test buildMapping  Time elapsed: 0 s  <<< FAILURE!
org.opentest4j.AssertionFailedError: expected: <true> but was: <false>
	at gov.cdc.prime.router.TranslatorTests.test buildMapping(TranslatorTests.kt:35)`
on this assertion:         ` assertEquals(true, oneToTwo.useDefault.contains("b"))`

2.  Unit test failed. 	at `gov.cdc.prime.router.SimpleReportTests.test producing az data from simplereport(SimpleReportTests.kt:100)`.  This is a test that tries to translate ./src/test/csv_test_files/input/simplereport.csv, to the az csv schema.   Errors:
- concat(Organization, Labname) failed - empty column in AZ data.  Expecting "Org1, Lab1"
- similarly, concat(Organization, ordering_facility_name) failed  - empty column in AZ data.  Expecting "Org1, Flat Plains"
- mapping from Result_ID column in simplereport.csv thru message_id -> placer_order_id -> testing_lab_specimen_id -> Specimen_ID column in AZ data is failing:  getting an empty column in AZ data.
- Test_code and Test_name column data empty in AZ data, expecting `94563-4` and `SARS-CoV-2 (COVID-19) Ag [Presence] in Respiratory specimen by Rapid immunoassay`, respectively.

3.   Formatting changes created during mvn build need to be checked in in Element.kt, Mappers.kt, etc.

NOTE:   ./quick-test.sh found the same csv schema generation problems as the above unit test, as well as a missing `Device_type` value in the Pima schema.

**Comment by RickHawesUSDS** [comment]
@jimduff-usds 
The build works now and the `./quick_start.sh` tests pass. One issue that should continue to be discussed was caused by the `device_type` field that was created for PIMA. It maps one type to another type. Pretty sure this behavior is "wrong" in one sense or another, but I left it in.

File: prime-router/src/main/kotlin/azure/DatabaseAccess.kt [hunk 23]
```diff
@@ -237,6 +241,14 @@ class DatabaseAccess(private val create: DSLContext) {
...
             config.addDataSourceProperty("prepStmtCacheSize", "250")
             config.addDataSourceProperty("prepStmtCacheSqlLimit", "2048")
+            // See this info why these are a good value
+            //  https://github.com/brettwooldridge/HikariCP/wiki/About-Pool-Sizing
+            config.minimumIdle = 2
+            config.maximumPoolSize = 8
+            // This strongly recommended to be set "be several seconds shorter than any database or infrastructure
+            // imposed connection time limit". Not sure what value is but have observed that connection are closed
+            // after about 10 minutes
+            config.maxLifetime = 180000
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/azure/DatabaseAccess.kt:251]: I've been meaning to check whether there's a simple way to just have connections stay open "forever" when I'm running in the debugger locally.    Connections closing make the debugging more confusing, especially if you are trying to debug something database related!

File: prime-router/src/main/kotlin/azure/DatabaseAccess.kt [hunk 24]
```diff
@@ -192,6 +193,7 @@ class DatabaseAccess(private val create: DSLContext) {
         currentAction: Event.Action,
         nextAction: Event.Action,
         nextActionAt: OffsetDateTime? = null,
+        retryToken: String? = null,
         txn: DataAccessTransaction,
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/azure/DatabaseAccess.kt:197]: Very glad to see the retry functionality!

File: prime-router/src/main/kotlin/serializers/RedoxSerializer.kt [hunk 25]
```diff
@@ -0,0 +1,181 @@
+package gov.cdc.prime.router.serializers
```
Comment (jimduff-usds) [inline prime-router/src/main/kotlin/serializers/RedoxSerializer.kt:1]: I did a quick sanity check - Generate some redox json, and did a quick eyeball sanity check in a prettyprinter.   All looks good so far.

**Review by jimduff-usds — APPROVED** [review]
Some minor comments.   
Let's merge in Mo's changes and I can re-test again.

**Pull request merged**